[LoopInterchange] Motivating example for interchange. NFC.

sjoerdmeijer opened the pull request:
This is precommitting a full reproducer of one of our motivating examples.  Looking at a full reproducer is helpful for further discussion on DependenceAnalysis and Delinearization issues and the runtime predicates discussion. I appreciate that this is a larger than usual test case, but that is by design, because I think it is useful to look at the whole thing with all of its complexities. 

I have given useful names to all the relevant loop variables, and the relevant blocks in these loops and their functions, but have intentionally not done that for others as there are quite a few more.

kasuga-fj commented on llvm/test/Transforms/LoopInterchange/large-nested-6d.ll:
```diff
@@ -0,0 +1,543 @@
+;        // Stmt 3
+;        // Stmt 4
+;      End
+;     End
```
Just to confirm, this is a Fortran-based pseudo code, right? If so, it would be helpful to clarify that the data is stored in column-major order.

kasuga-fj reviewed (commented):
Thanks for the reproducer. I'll take a closer look at the details. Just some initial quick comments.

kasuga-fj reviewed (commented):
There are at least two problems if we want to interchange this.

- Currently this is not even subject to LoopInterchange in the first place. This is because LoopInterchange doesn't support a loop that has multiple sub-loops (i.e., `L->getSubLoops().size() >= 2`). In this case, perhaps either `Stmt2`, `Stmt3` or `Stmt4` is a loop as well? We probably need to add a feature to limit the scope of what LoopInterchange handles, rather than applying it to the entire loop nest. I think implementing this itself would not be difficult, but the heuristic for choosing which loop nest to analyze may be non-trivial. Similar functionality will likely be necessary for DA as well.
- Probably DA cannot provide something useful without the information about the range of `NX` and `NY`. I think there are only two options here: 
    1. Use runtime checks
    1. Tell the knowledge of the semantics of the original language, e.g., out-of-bounds array access is UB in C/C++/Fortran. Note that this is different from emitting array size information from frontend. What we need here is `NX < 54` and `NY < 54`. More precisely, `%18 < 54` and `%20 < 54` in the LLVM IR. Metadata about array sizes won't guarantee them.

I don't know if there exists another approach and which one is the best...

sjoerdmeijer commented:
Yes, exactly, this won't get interchanged yet. I will therefore change the subject of this merge request because I don't want to suggest that transforming this case is a requirement for interchange default enablement. Interchanging this loop-nest is the end goal, not where we start. Before we get this interchanged, we need to solve the two areas that you mentioned, and that is then assuming DA is correct, so interchanging this is many steps down the road. 

What I will do is the following:
- Make the minor changes you suggested: the pseudo-code is indeed from fortran so column-major, that is a good point to clarify,
- Add the analysis why this loop-nest doesn't get interchanged as comments in the test-case,
- Add a second function, that only comprises of the perfectly nested sub part of this bigger loop nest, so a reduced test case.

Having both tests allows us to continue discussing the different problems and also see the full picture, so the full test-case is to discuss the perfect vs. imperfectly nesting, and the perfectly nested test to discuss DA issues. 

I started looking into the full reproducer yesterday, and came to the same conclusion as you regarding the two issues that you mentioned. Regarding the first one, the nesting, I don't have the answer yet if interchange should look for perfectly subparts in a bigger loop nest, or if we should distribute this first. But anyway, I will do what I wrote above, and will first pick up the second test-case that I mentioned, the one that is perfectly nested, because then we can connect the dots with runtime predicates and see if and how that is going to work for this case.

sjoerdmeijer committed:
[LoopInterchange] Motivating example for interchange. NFC.

This is precommitting a full reproducer of one of our motivating
examples. Looking at a full reproducer is helpful for further
discussions on DependenceAnalysis and Delinearization issues.

I have given useful names to all the relevant loop variables, and the
relevant blocks in these loops and their functions, but have
intentionally not done that for others as there are quite a few more.

sjoerdmeijer commented:
All comments have now been addressed in latest revision [c3025b9](https://github.com/llvm/llvm-project/commit/c3025b9b9acb193ce0d2a2bf8ebfb0441397c33c).

sebpop commented on llvm/test/Transforms/LoopInterchange/large-nested-4d.ll:
```diff
@@ -0,0 +1,155 @@
+;
+; This reproducer is useful to focus on only on the 2nd challenge: the data
+; dependence analysis problem, and not worry about the rest of loop nest
+; structure.
```
Thanks for reducing the testcase even further and focusing on only one aspect of the interchange at a time.

kasuga-fj reviewed (approved):
I'm not sure if it is a good idea to add motivating examples where interchange still cannot be applied to regression tests (if the remark outputs change frequently due to minor modifications to LoopInterchange, it might be burdensome), but I don't have any strong objections.
